macaroonbakery/oven.py

```diff
@@ -0,0 +1,239 @@
+        The macaroon will expire at the given time - a time_before first party
+        caveat will be added with that time.
+
+        @return:
+        '''
+        if len(ops) == 0:
+            raise ValueError('cannot mint a macaroon associated '
+                             'with no operations')
+
+        ops = canonical_ops(ops)
+        root_key, storage_id = self.root_keystore_for_ops(ops).root_key()
+
+        id = self.macaroon_id(ops, storage_id, expiry)
+
+        id_bytes = six.int2byte(bakery.LATEST_BAKERY_VERSION) + \
+            id.SerializeToString()
+
+        if macaroon.macaroon_version(version) < MACAROON_V2:
+            # The old macaroon format required valid text for the macaroon id,
+            # so base64-encode it.
+            id_bytes = utils.raw_urlsafe_b64encode(id_bytes)
+
+        m = macaroon.Macaroon(root_key, id_bytes, self.location, version, None)
+        m = self.add_caveat(m, checkers.time_before_caveat(expiry))
+        m = self.add_caveats(m, caveats)
+        return m
+
+    def add_caveat(self, m, cav):
+        '''Add a caveat to the given macaroon.
+        '''
+        return m.add_caveat(cav, self.key, self.locator)
+
+    def add_caveats(self, m, cavs):
+        '''Add all the caveats to the given macaroon.
+        '''
+        return m.add_caveats(cavs, self.key, self.locator)
+
+    def macaroon_id(self, ops, storage_id, expiry):
+        nonce = os.urandom(16)
+        if len(ops) == 1 or self.ops_store is None:
+            return id_pb2.MacaroonId(
+                nonce=nonce,
+                storageId=storage_id,
+                ops=_macaroon_id_ops(ops))
+        # We've got several operations and a multi-op store, so use the store.
+        # TODO use the store only if the encoded macaroon id exceeds some size?
+        entity = self.ops_entity(ops)
+        self.ops_store.put_opts(entity, ops, expiry)
+        return id_pb2.MacaroonId(
+            nonce=nonce,
+            storageId=storage_id,
+            ops=[id_pb2.Op(entity=entity, actions=["*"])])
+
+    def ops_entity(self, ops):
+        ''' Returns a new multi-op entity name string that represents
+        all the given operations and caveats. It returns the same value
+        regardless of the ordering of the operations. It assumes that the
+        operations have been canonicalized and that there's at least one
+        operation.
+
+        :param ops:
+        :return: string that represents all the given operations and caveats.
+        '''
+        # Hash the operations, removing duplicates as we go.
+        hash_entity = hashlib.sha256()
+        for op in ops:
+            hash_entity.update('{}\n{}\n'.format(
+                op.action, op.entity).encode())
+        hash_encoded = base64.urlsafe_b64encode(hash_entity.digest())
+        return 'multi-' + hash_encoded.decode('utf-8').rstrip('=')
+
+    def macaroon_ops(self, macaroons):  # (ops []Op, conditions []string) {
```

Comment:
What's the comment here for?

Reply:
We should probably comment that this method makes the oven satisfy the MacaroonOpStore protocol required by the Checker class.

Reply:
done.